src/ray/raylet/node_manager.cc

```diff
@@ -844,13 +844,36 @@ void NodeManager::ProcessClientMessage(
 void NodeManager::ProcessRegisterClientRequestMessage(
     const std::shared_ptr<LocalClientConnection> &client, const uint8_t *message_data) {
   auto message = flatbuffers::GetRoot<protocol::RegisterClientRequest>(message_data);
-  client->SetClientID(from_flatbuf<ClientID>(*message->worker_id()));
+  auto client_id = from_flatbuf<ClientID>(*message->worker_id());
+  client->SetClientID(client_id);
   auto worker = std::make_shared<Worker>(message->worker_pid(), message->language(),
                                          message->port(), client);
   if (message->is_worker()) {
     // Register the new worker.
+    bool use_push_task = (worker->Port() > 0);
+    if (use_push_task) {
+      auto entry = worker_clients_.find(client_id);
+      if (entry != worker_clients_.end()) {
+        RAY_LOG(WARNING) << "Received registration of a new worker that already exists: "
+                         << client_id;
+        return;
+      }
+
+      // Initialize a rpc client to the new worker.
+      std::unique_ptr<rpc::WorkerTaskClient> grpc_client(
+          new rpc::WorkerTaskClient("127.0.0.1", worker->Port(), client_call_manager_));
+      worker_clients_.emplace(client_id, std::move(grpc_client));
+    }
+
     worker_pool_.RegisterWorker(std::move(worker));
-    DispatchTasks(local_queues_.GetReadyTasksWithResources());
+    if (use_push_task) {
+      auto registered_worker = worker_pool_.GetRegisteredWorker(client);
+      HandleWorkerAvailable(registered_worker);
```

Comment:
Can you leave a comment explaining why this is only necessary for the `use_push_task` case?

Reply:
done